### ISSUE TITLE
Remote Pollers sometimes run until killed

### ISSUE THREAD

**Issue by ctrowat** [Cacti/cacti#4468]
# Describe the bug

We run Cacti with 4 remote pollers.  They are each configured for 20 processes with 2 threads and are running spine in VMs with 2 vCPUs and 4GB of RAM.  Seemingly randomly each of them will either finish in 20 seconds, or run until the time limit expires.  I am having trouble deciphering whether this is a spine issue or a cacti issue.  That is to say, whether the problem is with what spine is doing or with how cacti manages the threads.

When running the pollers on the command line with debug enabled we get output like this:

```
2021-11-19 23:38:19 - POLLER: Poller[2] PID[23963] NOTE: Poller Int: '60', Cron Int: '60', Time Since Last: '28.28', Max Runtime '58', Poller Runs: '1'
2021-11-19 23:38:19 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=6 --last=170 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=171 --last=672 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=873 --last=2346 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=2347 --last=2473 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=2474 --last=2566 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=2567 --last=2690 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=2691 --last=2803 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=2804 --last=4900 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=4901 --last=4970 --mibs]
2021-11-19 23:38:20 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=4971 --last=5036 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=5037 --last=5131 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=5132 --last=5881 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=5931 --last=9398 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=9399 --last=9446 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=9447 --last=10047 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=10048 --last=10076 --mibs]
2021-11-19 23:38:21 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=10077 --last=10128 --mibs]
2021-11-19 23:38:22 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /opt/cacti/spine, ARGS:  --mode=online --poller=2 --first=10129 --last=12236 --mibs]
Waiting on 18 of 18 pollers.
Waiting on 18 of 18 pollers.
Waiting on 18 of 18 pollers.
Waiting on 18 of 18 pollers.
Waiting on 18 of 18 pollers.
Waiting on 18 of 18 pollers.
Waiting on 17 of 18 pollers.
Waiting on 15 of 18 pollers.
Waiting on 13 of 18 pollers.
Waiting on 13 of 18 pollers.
Waiting on 11 of 18 pollers.
Waiting on 9 of 18 pollers.
Waiting on 7 of 18 pollers.
Waiting on 6 of 18 pollers.
Waiting on 5 of 18 pollers.
Waiting on 3 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
Waiting on 2 of 18 pollers.
2021-11-19 23:39:18 - POLLER: Poller[2] PID[23963] Maximum runtime of 58 seconds exceeded. Exiting.
2021-11-19 23:39:18 - SYSTEM STATS: Time:58.4006 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:39:19 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /usr/local/bin/php, ARGS:  -q '/opt/cacti/cacti/poller_automation.php' -M]
2021-11-19 23:39:19 - POLLER: Poller[2] PID[23963] DEBUG: About to Spawn a Remote Process [CMD: /usr/local/bin/php, ARGS:  -q '/opt/cacti/cacti/poller_maintenance.php']
```

However this is very interesting because we only get 16 rows for this poller in the database despite it clearly spawning 18 threads at the top!

![image](https://user-images.githubusercontent.com/9736285/142704626-4237f3f8-e135-497f-9ca6-299146199bfc.png)

## To Reproduce

Honestly I don't know why our pollers started doing this so I couldn't tell you exactly how to reproduce this.

## Expected behavior

Poller runtime should be somewhat stable and all poller processes should be accounted for in the database.

## Extra Information

Here is what we are seeing in the remote poller's logs:

```
2021-11-19 23:28:01 - SYSTEM STATS: Time:58.4596 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
--
2021-11-19 23:28:01 - SYSTEM WARNING: Primary Admin account notifications disabled! Unable to send administrative Email.
2021-11-19 23:28:01 - POLLER: Poller[Cacti Poller 01] PID[8621] Maximum runtime of 58 seconds exceeded. Exiting.
2021-11-19 23:26:21 - SYSTEM STATS: Time:19.4611 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:26:00 - SYSTEM STATS: Time:58.4897 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:26:00 - SYSTEM WARNING: Primary Admin account notifications disabled! Unable to send administrative Email.
2021-11-19 23:26:00 - POLLER: Poller[Cacti Poller 01] PID[6082] Maximum runtime of 58 seconds exceeded. Exiting.
2021-11-19 23:25:00 - SYSTEM STATS: Time:58.3223 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:25:00 - SYSTEM WARNING: Primary Admin account notifications disabled! Unable to send administrative Email.
2021-11-19 23:25:00 - POLLER: Poller[Cacti Poller 01] PID[4843] Maximum runtime of 58 seconds exceeded. Exiting.
2021-11-19 23:23:22 - SYSTEM STATS: Time:19.4759 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:22:20 - SYSTEM STATS: Time:18.3579 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:21:21 - SYSTEM STATS: Time:19.4182 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
2021-11-19 23:21:00 - SYSTEM STATS: Time:58.2859 Method:spine Processes:20 Threads:2 Hosts:1233 HostsPerProcess:62 DataSources:16286 RRDsProcessed:0
```

**Comment by netniV** [comment]
I think this is a duplicate of #4460, if so, please close.  If not, please give an explanation as why it is different.

**Comment by ctrowat** [comment]
This is a different issue - I first had the other problem then the provided change set resolved that part and now I have this issue (or more likely, had this issue all along but it was masked by the other one).  I am not sure whether the problem is poller.php not correctly starting spine, or whether spine is failing before inserting the necessary rows in the poller_time table but the end result is poller.php gets stuck in the sleep part of the loop waiting for threads that never got placed in the database to complete because it's checking an ideal number of pollers against the number in the database that have completed.

**Comment by netniV** [comment]
You may have some overrunning processes, to find that out, you would need to increase your logging level of spine.  Each of those spine commands that you have are separating out a chunk of hosts each.  Try running each one in readonly and increase logging to see if you can narrow down which is causing you the issue.

One of the things I tried to introduce was a timeout of certain scripts, but that caused other issues so was reverted, but the overall processing timeout remains so if a script does overrun, it will affect others.

**Comment by ctrowat** [comment]
Thanks for the reply!

Inspecting the process list in the container shows all of the spine processes end with lots of time left over and there are no overrunning spine processes.

Run individually each spine command runs properly and ends properly, and it doesn't happen 100% of the time when they run concurrently, I have yet to identify the cause.

My current thought is that the spine processes are somehow crashing before inserting themselves into the database I just haven't had time to instrument the processes to inspect what is going on in more detail as I have been moved to another project for the next little while.

**Comment by TheWitness** [comment]
This one is resolved now in the 1.2.x branch and is duplicate of another report.  Just update `lib/poller.php` from the 1.2.x branch.  Thanks!
